src/SamplesApp/SamplesApp.UITests/TestFramework/ImageAssert.cs

```diff
@@ -57,57 +54,71 @@ public static void AreEqual(FileInfo expected, Rectangle expectedRect, FileInfo
 				for (var x = 0; x < width; x++)
 					for (var y = 0; y < height; y++)
 					{
-						Assert.AreEqual(
-							expectedBitmap.GetPixel(x + expectedOffset.x, y + expectedOffset.y),
-							actualBitmap.GetPixel(x + actualOffset.x, y + actualOffset.y),
-							$"Pixel {x},{y}");
+						var expectedColor = expectedBitmap.GetPixel(x + expectedOffset.x, y + expectedOffset.y);
+						var actualColor = actualBitmap.GetPixel(x + actualOffset.x, y + actualOffset.y);
+
+						if (expectedColor != actualColor)
+						{
+							Assert.Fail(WithContext(builder: builder => builder
+								.AppendLine($"Pixels at ({x}, {y}) are not the same")
+								.AppendLine($"expectedColor: {ToArgbCode(expectedColor)} {expectedColor}")
+								.AppendLine($"actualColor  : {ToArgbCode(actualColor)} {actualColor}")
+							));
+						}
 					}
 			}
+
+			string WithContext(string message = null, Action<StringBuilder> builder = null)
+			{
+				return new StringBuilder()
+					.AppendLine($"ImageAssert.AreEqual @ line {line}")
+					.AppendLine("expected: " + expected?.Name + (expectedRect == FirstQuadrant ? null : $"in {expectedRect}"))
+					.AppendLine("actual  : " + actual?.Name + (actualRect == FirstQuadrant ? null : $"in {actualRect}"))
+					.AppendLine("====================")
+					.ApplyIf(message != null, x => x.AppendLine(message))
+					.ApplyIf(builder != null, builder)
+					.ToString();
+			}
 		}
+		#endregion
 
+		#region AreAlmostEqual
 		/// <summary>
 		/// Asserts that two screenshots are equal to each other inside the area of the nominated rect to within the given pixel error
 		/// (ie, the A, R, G, or B values cumulatively differ by more than the permitted error for any pixel).
 		/// </summary>
-		public static void AreAlmostEqual(FileInfo expected, FileInfo actual, IAppRect rect, int permittedPixelError)
-			=> AreAlmostEqual(expected, actual, new Rectangle((int)rect.X, (int)rect.Y, (int)rect.Width, (int)rect.Height), permittedPixelError);
+		public static void AreAlmostEqual(FileInfo expected, FileInfo actual, IAppRect rect, int permittedPixelError, [CallerLineNumber] int line = 0)
+			=> AreAlmostEqualImpl(expected, rect.ToRectangle(), actual, rect.ToRectangle(), permittedPixelError, line);
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Refactor this method to add validation of parameter 'rect' before using it.](https://app.codacy.com/gh/unoplatform/uno/pullRequest?prid=5170992)